lib/spack/llnl/util/tty/log.py

```diff
@@ -88,43 +129,122 @@ def __init__(self, stream):
         will do nothing.
         """
         self.stream = stream
+        self.signals = [signal.SIGALRM, signal.SIGTSTP, signal.SIGCONT]
+
+    def _is_background(self):
+        """True iff calling process is in the background."""
+        return _is_background_tty(self.stream)
+
+    def _get_canon_echo_flags(self):
+        """Get current termios canonical and echo settings."""
+        cfg = termios.tcgetattr(self.stream)
+        return (
+            bool(cfg[3] & termios.ICANON),
+            bool(cfg[3] & termios.ECHO),
+        )
+
+    def _enable_keyboard_input(self):
+        """Disable canonical input and echoing on ``self.stream``."""
+        # "enable" input by disabling canonical mode and echo
+        new_cfg = termios.tcgetattr(self.stream)
+        new_cfg[3] &= ~termios.ICANON
+        new_cfg[3] &= ~termios.ECHO
+
+        # Apply new settings for terminal
+        with ignore_signal(signal.SIGTTOU):
+            termios.tcsetattr(self.stream, termios.TCSANOW, new_cfg)
+
+    def _restore_input(self):
+        """Restore the original input configuration on ``self.stream``."""
+        # _restore_input Can be called in foreground or background. When called
+        # in the background, tcsetattr triggers SIGTTOU, which we must ignore,
+        # or the process will be stopped.
+        with ignore_signal(signal.SIGTTOU):
+            termios.tcsetattr(self.stream, termios.TCSANOW, self.old_cfg)
+
+    @contextmanager
+    def defer_signals(self):
+        with defer_signals(self.signals):
+            yield
+
+    def _bg_fg_handler(self, signum, frame):
+        """Ensures that input is enabled and disabled at the right times.
+
+        We want canonical input and echo to be disabled when the calling
+        process group is in the foreground, and disabled when it is in
+        the background. This handler
+
+        This is the handler for ``SIGTSTP``, ``SIGCONT``, and ``SIGALRM``.
+
+        """
+        with self.defer_signals():
+            # query terminal flags and fg/bg status
+            flags = self._get_canon_echo_flags()
+            bg = self._is_background()
+
+            # restore sanity if any flags are amiss
+            if not bg and any(flags):    # fg, but input not enabled
```

Comment:
Can you add comments that refer to the excellent state diagram above (e.g. this corresponds to `SIGALRM` handling)?

Reply:
It's a bit more complicated now that everything uses the same handler -- the same calls are now used for SIGTSTP and SIGCONT -- so this isn't just for SIGALRM.

I added "see diagram in class docs"

Reply:
I think in general everything may not map 1-to-1, although my understanding was that this particular block (inside the if) was for handling SIGALRM: based on the diagram, I wouldn't expect to be here otherwise.

Reply:
This is outdated -- no more SIGALRM